server/src/test/java/org/apache/kafka/server/AssignmentsManagerTest.java

```diff
@@ -14,451 +14,506 @@
+    static class TestEnv implements AutoCloseable {
+        final ExponentialBackoff backoff;
+        final MockNodeToControllerChannelManager channelManager;
+        final MetricsRegistry metricsRegistry = new MetricsRegistry();
+        final AssignmentsManager assignmentsManager;
+        final Map<TopicIdPartition, Integer> successes;
+
+        TestEnv() {
+            this.backoff = new ExponentialBackoff(1, 2, 4, 0);
+            this.channelManager = new MockNodeToControllerChannelManager();
+            this.assignmentsManager = new AssignmentsManager(
+                    backoff, Time.SYSTEM, channelManager, 1, () -> TEST_IMAGE, metricsRegistry);
+            this.successes = new HashMap<>();
+        }
+
+        void onAssignment(TopicIdPartition topicIdPartition, Uuid directoryId) {
+            assignmentsManager.onAssignment(topicIdPartition, directoryId, "test", () -> {
+                synchronized (successes) {
+                    successes.put(topicIdPartition, successes.getOrDefault(topicIdPartition, 0) + 1);
+                }
+            });
+        }
+
+        int success(TopicIdPartition topicIdPartition) {
+            synchronized (successes) {
+                return successes.getOrDefault(topicIdPartition, 0);
             }
-            if (readyToAssert.getCount() == 3) {
-                invocation.getArgument(1, ControllerRequestCompletionHandler.class).onComplete(
-                    new ClientResponse(null, null, null, 0L, 0L, false, false,
-                        new UnsupportedVersionException("test unsupported version exception"), null, null));
+        }
 
-                // duplicate should be ignored
-                manager.onAssignment(new TopicIdPartition(TOPIC_1, 2), DIR_3, "testRequeuesFailedAssignmentPropagations", () -> { });
+        void successfullyCompleteCallbackOfRequestAssigningTopic1ToDir1() throws Exception {
+            channelManager.completeCallback(req -> {
+                AssignReplicasToDirsRequestData.DirectoryData directoryData = req.directories().get(0);
+                assertEquals(DIR_1, directoryData.id());
+                AssignReplicasToDirsRequestData.TopicData topicData = directoryData.topics().get(0);
+                assertEquals(TOPIC_1, topicData.topicId());
+                assertEquals(0, topicData.partitions().get(0).partitionIndex());
+                return mockClientResponse(new AssignReplicasToDirsResponseData().
+                    setDirectories(Arrays.asList(new AssignReplicasToDirsResponseData.DirectoryData().
+                        setId(DIR_1).
+                        setTopics(Arrays.asList(new AssignReplicasToDirsResponseData.TopicData().
+                            setTopicId(TOPIC_1).
+                            setPartitions(Arrays.asList(new AssignReplicasToDirsResponseData.PartitionData().
+                                setPartitionIndex(0).
+                                setErrorCode((short) 0))))))));
```

Comment:
There's `org.apache.kafka.metadata.AssignmentsHelper#buildResponseData` which can simplify this.

Reply:
hmm, I don't think that would be fewer lines since I'd have to create the nested maps